rocprim/include/rocprim/device/detail/device_merge.hpp

```diff
@@ -72,28 +72,24 @@ range_t compute_range(const unsigned int id,
     return range_t{p1, p2, diag1 - p1, diag2 - p2};
 }
 
-template<
-    class KeysInputIterator1,
-    class KeysInputIterator2,
-    class BinaryFunction
->
+template <class KeysInputIterator, class OffsetT, class BinaryFunction>
 ROCPRIM_DEVICE ROCPRIM_INLINE
-unsigned int merge_path(KeysInputIterator1 keys_input1,
-                        KeysInputIterator2 keys_input2,
-                        const size_t input1_size,
-                        const size_t input2_size,
-                        const unsigned int diag,
-                        BinaryFunction compare_function)
```

Comment:
This change breaks calls to `thrust::merge` (which allows separate iterators). In particular, [here](https://github.com/ROCmSoftwarePlatform/rocThrust/blob/develop/thrust/system/hip/detail/merge.h#L101-L111) is where rocthrust calls `rocprim::merge` and [here](https://github.com/ROCmSoftwarePlatform/rocPRIM/blob/develop/rocprim/include/rocprim/device/detail/device_merge.hpp#L56-L88) is where `partition_kernel_impl` (which accepts distinct iterators) calls `merge_path`, which now does not. This affects PETSc with ROCm-5.3.

<details>

```
In file included from /projects/petsc/src/mat/impls/aij/seq/seqhipsparse/aijhipsparse.hip.cpp:15:
In file included from /projects/petsc/include/../src/mat/impls/aij/seq/seqhipsparse/hipsparsematimpl.h:7:
In file included from /projects/petsc/include/../src/vec/vec/impls/seq/cupm/vecseqcupm.hpp:13:
In file included from /opt/rocm-5.3.0/include/thrust/transform_reduce.h:194:
In file included from /opt/rocm-5.3.0/include/thrust/detail/transform_reduce.inl:25:
In file included from /opt/rocm-5.3.0/include/thrust/system/detail/generic/transform_reduce.h:51:
In file included from /opt/rocm-5.3.0/include/thrust/system/detail/generic/transform_reduce.inl:19:
In file included from /opt/rocm-5.3.0/include/thrust/reduce.h:781:
In file included from /opt/rocm-5.3.0/include/thrust/detail/reduce.inl:28:
In file included from /opt/rocm-5.3.0/include/thrust/system/detail/generic/reduce_by_key.h:87:
In file included from /opt/rocm-5.3.0/include/thrust/system/detail/generic/reduce_by_key.inl:37:
In file included from /opt/rocm-5.3.0/include/thrust/scan.h:1656:
In file included from /opt/rocm-5.3.0/include/thrust/detail/scan.inl:28:
In file included from /opt/rocm-5.3.0/include/thrust/system/detail/adl/scan.h:44:
In file included from /opt/rocm-5.3.0/include/thrust/system/hip/detail/scan.h:48:
In file included from /opt/rocm-5.3.0/include/rocprim/rocprim.hpp:55:
In file included from /opt/rocm-5.3.0/include/rocprim/device/device_merge.hpp:31:
/opt/rocm-5.3.0/include/rocprim/device/detail/device_merge.hpp:133:9: error: no matching function for call to 'merge_path'
        merge_path(
        ^~~~~~~~~~
/opt/rocm-5.3.0/include/rocprim/device/device_merge.hpp:57:5: note: in instantiation of function template specialization 'rocprim::detail::partition_kernel_impl<unsigned int *, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<dou
ble>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, IJCompare4>' requested here
    partition_kernel_impl(
    ^
/opt/rocm-5.3.0/include/rocprim/device/device_merge.hpp:180:33: note: in instantiation of function template specialization 'rocprim::detail::partition_kernel<unsigned int *, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double
>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, IJCompare4>' requested here
        HIP_KERNEL_NAME(detail::partition_kernel),
                                ^
/opt/rocm-5.3.0/include/rocprim/device/device_merge.hpp:299:20: note: in instantiation of function template specialization 'rocprim::detail::merge_impl<rocprim::default_config, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<dou
ble>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_it
erator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>>>, rocprim::empty_type *, rocprim::empty_type *, rocprim::empty_type *, IJCompare4>' requested here
    return detail::merge_impl<Config>(
                   ^
/opt/rocm-5.3.0/include/thrust/system/hip/detail/merge.h:101:46: note: in instantiation of function template specialization 'rocprim::merge<rocprim::default_config, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrus
t::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrus
t::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>>>, IJCompare4>' requested here
        hip_rocprim::throw_on_error(rocprim::merge(NULL,
                                             ^
/opt/rocm-5.3.0/include/thrust/system/hip/detail/merge.h:258:23: note: in instantiation of function template specialization 'thrust::hip_rocprim::__merge::merge<thrust::hip_rocprim::par_t, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::de
vice_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detai
l::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>>>, IJCompare4>' requested here
      return __merge::merge(
                      ^
/opt/rocm-5.3.0/include/thrust/system/hip/detail/merge.h:242:10: note: in instantiation of member function 'thrust::hip_rocprim::merge(execution_policy<thrust::hip_rocprim::par_t> &, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_p
tr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int
>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>>>, thru
st::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>>>, IJCompare4)::workar
ound::par' requested here
  struct workaround
         ^
/opt/rocm-5.3.0/include/thrust/detail/merge.inl:65:10: note: in instantiation of function template specialization 'thrust::hip_rocprim::merge<thrust::hip_rocprim::par_t, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>,
thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<
thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>>>, IJCompare4>' requested here
  return merge(thrust::detail::derived_cast(thrust::detail::strip_const(exec)), first1, last1, first2, last2, result, comp);
         ^
/projects/petsc/src/mat/impls/aij/seq/seqhipsparse/aijhipsparse.hip.cpp:4625:31: note: in instantiation of function template specialization 'thrust::merge<thrust::hip_rocprim::par_t, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_p
tr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thrust::tuple<thrust::detail::nor
mal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>>>, IJCompare4>' requested here
      PetscCallThrust(thrust::merge(thrust::device, Azb, Aze, Bzb, Bze, Czb, IJCompare4()));
                              ^
/opt/rocm-5.3.0/include/rocprim/device/detail/device_merge.hpp:77:9: note: candidate template ignored: deduced conflicting types for parameter 'KeysInputIterator' ('zip_iterator<tuple<[...], thrust::detail::normal_iterator<thrust::device_ptr<int>>, [8 * ...]>>' vs. 'zip_iterator<tuple<[...], thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>
>>, [8 * ...]>>')
OffsetT merge_path(KeysInputIterator keys_input1,
        ^
```
</details>

Cc: @suyashtn @jczhang07

Reply:
@jychang48 :wave: You also worked on this MR for PETSc, that is now broken after this change in rocprim.

Reply:
@jedbrown yes I also noticed this break. l’ll look into it

Reply:
Can you confirm if https://github.com/ROCmSoftwarePlatform/rocPRIM/pull/381 solves this issue?

Reply:
It looks like it should work, but how does one test it within a (Debian) install of rocm-5.3.0?

Reply:
@nolmoonen  #381 was suggested to fix the errors @jedbrown is reporting here... The changes in #381 basically reverts the API change and works.

Reply:
Great, how can we get this fix shipped to users? More people will attempt to install PETSc with rocm-5.3 over the coming months (and I'm sure we're not the only ones). I think the best case is that 5.3.1 ships soon with this fix and we just encourage OLCF and other facilities apply it. Otherwise, we need to make a work-around (could be install rocPRIM from source in PETSc or work around the limitation in code; both are a bit of a maintenance headache).

Reply:
Given that this is a breaking issue, I **might** be able to get #381 cherry-picked into 5.3.2...anything earlier than that, the ship has sailed, unfortunately.  But no promises, let me confer with @doctorcolinsmith on this and get back to you.